# Exploración

Comenzamos cargando las librerías


In [1]:
import sys, platform
import pandas as pd
import numpy as np

print("Python", sys.version)
print("Platform", platform.platform())
print("pandas:", pd.__version__)
print("numpy:", np.__version__)

Python 3.12.10 (tags/v3.12.10:0cc8128, Apr  8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)]
Platform Windows-11-10.0.26100-SP0
pandas: 2.3.2
numpy: 2.3.3


Cargamos los datos

In [2]:
data = "C:/Users/hered/Desktop/TFM/TFM/data/PMC-Patients.csv"

df = pd.read_csv(data, dtype=str, encoding="utf-8", na_filter=False)

print("Filas:", len(df))
print("Columnas:", list(df.columns))

#visualizamos

try:
    display(df.head(3))
except NameError:
    print(df.head(3))

Filas: 167034
Columnas: ['patient_id', 'patient_uid', 'PMID', 'file_path', 'title', 'patient', 'age', 'gender', 'relevant_articles', 'similar_patients']


patient_id patient_uid      PMID                         file_path  \
0          0   7665777-1  33492400  comm/PMC007xxxxxx/PMC7665777.xml   
1          1   7665777-2  33492400  comm/PMC007xxxxxx/PMC7665777.xml   
2          2   7665777-3  33492400  comm/PMC007xxxxxx/PMC7665777.xml   

                                               title  \
0  Early Physical Therapist Interventions for Pat...   
1  Early Physical Therapist Interventions for Pat...   
2  Early Physical Therapist Interventions for Pat...   

                                             patient               age gender  \
0  This 60-year-old male was hospitalized due to ...  [[60.0, 'year']]      M   
1  A 39-year-old man was hospitalized due to an i...  [[39.0, 'year']]      M   
2  One week after a positive COVID-19 result this...  [[57.0, 'year']]      M   

                                   relevant_articles  \
0  {'32320506': 1, '32293716': 1, '23219649': 1, ...   
1  {'32320506': 1, '32293716': 1, '23219649': 1, ...   
2  {'32320506': 1, '32293716': 1, '23219649': 1, ...   

                                    similar_patients  
0  {'7665777-2': 2, '7665777-3': 2, '7665777-4': ...  
1  {'7665777-1': 2, '7665777-3': 2, '7665777-4': ...  
2  {'7665777-1': 2, '7665777-2': 2, '7665777-4': ...

Como exploración inicial, comenzamos viendo la distribución de gender y longitudes de la columna de patient

In [3]:
if "gender" in df.columns:
    print("Value conts de 'gender'")
    print(df["gender"].value_counts().head(10))
else:
    print("no existe")

Value conts de 'gender'
gender
M    87659
F    79375
Name: count, dtype: int64


In [4]:
if "patient" in df.columns:
    text_len = df["patient"].str.len()
    print("\nLongitud 'patient' - min/median/mean/max")
    print(int(text_len.min()), int(text_len.median()), round(text_len.mean(),1), int(text_len.max()))
else:
    print("no existe patient")


Longitud 'patient' - min/median/mean/max
55 2491 2765.3 95400


In [5]:
print("\nEjemplos de 'patient'")
for s in df.loc[df["patient"].str.len()>0, "patient"].head(2):
    print("-"*80)
    print(s.replace("\n"," "))


Ejemplos de 'patient'
--------------------------------------------------------------------------------
This 60-year-old male was hospitalized due to moderate ARDS from COVID-19 with symptoms of fever, dry cough, and dyspnea. We encountered several difficulties during physical therapy on the acute ward. First, any change of position or deep breathing triggered coughing attacks that induced oxygen desaturation and dyspnea. To avoid rapid deterioration and respiratory failure, we instructed and performed position changes very slowly and step-by-step. In this way, a position change to the 135° prone position () took around 30 minutes. This approach was well tolerated and increased oxygen saturation, for example, on day 5 with 6 L/min of oxygen from 93% to 97%. Second, we had to adapt the breathing exercises to avoid prolonged coughing and oxygen desaturation. Accordingly, we instructed the patient to stop every deep breath before the need to cough and to hold inspiration for better air di

#### Ejemplo del primer "patient" de nuestro dataset:

This 60-year-old male was hospitalized due to moderate ARDS from COVID-19 with symptoms of fever, dry cough, and dyspnea. We encountered several difficulties during physical therapy on the acute ward. First, any change of position or deep breathing triggered coughing attacks that induced oxygen desaturation and dyspnea. To avoid rapid deterioration and respiratory failure, we instructed and performed position changes very slowly and step-by-step. In this way, a position change to the 135° prone position () took around 30 minutes. This approach was well tolerated and increased oxygen saturation, for example, on day 5 with 6 L/min of oxygen from 93% to 97%. Second, we had to adapt the breathing exercises to avoid prolonged coughing and oxygen desaturation. Accordingly, we instructed the patient to stop every deep breath before the need to cough and to hold inspiration for better air distribution. In this manner, the patient performed the breathing exercises well and managed to increase his oxygen saturation. Third, the patient had difficulty maintaining sufficient oxygen saturation during physical activity. However, with close monitoring and frequent breaks, he managed to perform strength and walking exercises at a low level without any significant deoxygenation. Exercise progression was low on days 1 to 5, but then increased daily until hospital discharge to a rehabilitation clinic on day 10.

#### Segundo ejemplo:

A 39-year-old man was hospitalized due to an increasingly reduced general health condition, after persistent fever and dry cough for 2 weeks. The patient initially needed 4 L/min of oxygen, had a rapid and shallow breathing pattern at rest and became severely breathless during minor physical activities. In the beginning, physical therapy focused on patient education about dyspnea-relieving positions, the importance of regular mobilization, and deep-breathing exercises. However, it quickly became evident that his anxiety from fear of dying and worries about his future aggravated his dyspnea and vice versa. The patient was so dyspneic, anxious, and weak that he was barely able to walk to the toilet. To counter this vicious circle, the physical therapist actively listened to the patient, explained why he was experiencing breathlessness, and tested suitable positions to relieve his dyspnea. He seemed to benefit from the education and the relaxing breathing exercises, as seen on day 2, when his respiratory rate could be reduced from 30 breaths/min to 22 breaths/min and his oxygen saturation increased from 92% to 96% on 4 L/min oxygen after guiding him through some deep-breathing exercises. Over the next days, his dyspnea and anxiety started to alleviate and he regained his self-confidence. Therapy was progressively shifted to walking and strength training and the patient rapidly advanced to walk 350 m without a walking aid or supplemental oxygen before his discharge home.



Como hemos podido observar "age" viene como [[60.0, 'year']] y "gender" como M/F.  

El siguiente paso será normalizar "age" y "gender":

- age_col_norm (int) 
     
    "[[60.0, 'year']]"  -> 60
    "60" / "60.0"       -> 60
    Devuelve int o None si no encuentra un valor razonable (0<edad<130).

- sex_col_norm (male/female)

    Mapea variantes a {'male','female'}.
    Acepta: 'M','F','male','female','man','woman','hombre','mujer', etc.

El obejtivo es tener referecnias limpias a partir de los que ya viene estructurado en el CSV, para poder comparar contra lo que se extraiga del texto libre de la oclumna 'patient'.

In [18]:
import re

def parse_age_col(val:str):
    if not val or not isinstance(val, str):
        return None
    m = re.search(r'([0-9]+(?:\.[0-9]+)?)',val)
    if not m:
        return None
    try:
        num = float(m.group(1))
        if 0 < num < 130:
            return int(round(num))
    except:
        return None
    return None

def normalize_gender_col(val:str):
    if not val:
        return None
    v = val.strip().lower()
    if v in {"m", "male", "man", "hombre"}:
        return "male"
    if v in {"f", "female", "woman", "mujer"}:
        return "female"
    return None

#Creamos las columnas normalizadas:

df["age_col_norm"] = pd.to_numeric(df["age_col_norm"], errors="coerce").astype("Int64")
df["sex_col_norm"] = df.get("gender", "").apply(normalize_gender_col)

#Vista
try:
    display(df[["age", "age_col_norm", "gender", "sex_col_norm"]].head(5))
except NameError:
    print(df[["age","age_col_norm","gender","sex_col_norm"]].head(5))

age  age_col_norm gender sex_col_norm
0  [[60.0, 'year']]            60      M         male
1  [[39.0, 'year']]            39      M         male
2  [[57.0, 'year']]            57      M         male
3  [[69.0, 'year']]            69      M         male
4  [[57.0, 'year']]            57      M         male

El siguiente paso será extraer edad y sexo desde la columna "patient" y vamos a compararlas con las columnas normalizadas

In [22]:
# %% [markdown]
# CELDA 6 — ANTES
# Objetivo: Definir las regex y la función `extract_age_sex_from_text(note_text)`.
# Esta celda no depende de N_SAMPLE y la usaremos en TODAS las filas.

# %%
import re

# Patrones básicos que aparecen en tus textos:
# - "60-year-old", "39-year-old"
# - "sex: male", "gender: female", etc. (incluye variantes en español)
age_re_1 = re.compile(r'(?i)\b([0-9]{1,3})-year-old\b')
age_re_2 = re.compile(r'(?i)\bage\s*[:=]?\s*([0-9]{1,3})\b')
sex_re   = re.compile(r'(?i)\b(sex|gender)\s*[:=]?\s*(male|female|man|woman|hombre|mujer|m|f)\b')

def extract_age_sex_from_text(note_text: str):
    """
    Devuelve dict con:
      - age_from_text (int o None)
      - sex_from_text (string original o None)
      - sex_from_text_norm ('male'/'female' o None)
    Usa tu normalize_gender_col() para normalizar el sexo.
    """
    t = (note_text or "").replace("\u2009"," ").replace("\u00a0"," ")
    out = {}

    # Edad (primero "X-year-old", si no, "age: X")
    m = age_re_1.search(t) or age_re_2.search(t)
    if m:
        try:
            out["age_from_text"] = int(m.group(1))
        except:
            out["age_from_text"] = None
    else:
        out["age_from_text"] = None

    # Sexo (normalizamos a male/female con tu función)
    m = sex_re.search(t)
    if m:
        out["sex_from_text"] = m.group(2).lower()
        out["sex_from_text_norm"] = normalize_gender_col(out["sex_from_text"])
    else:
        out["sex_from_text"] = None
        out["sex_from_text_norm"] = None

    return out

# Prueba rápida opcional en 2 filas (para ver formato del dict):
try:
    for txt in df["patient"].head(2):
        print(extract_age_sex_from_text(txt))
except Exception as e:
    print("Prueba rápida: omitiendo por error:", e)

# %% [markdown]
# CELDA 6 — DESPUÉS
# Ya existe `extract_age_sex_from_text`. Ahora puedes ejecutar tu CELDA 7 para procesar todas las filas.


{'age_from_text': 60, 'sex_from_text': None, 'sex_from_text_norm': None}
{'age_from_text': 39, 'sex_from_text': None, 'sex_from_text_norm': None}


In [20]:
# %% [markdown]
# CELDA 7A — ANTES
# Objetivo: Mejorar la extracción de SEXO desde el texto `patient`.
# - Capturamos: "60-year-old male", "A 39-year-old man", "female patient", etc.
# - Evitamos "female-to-male" con lookarounds que descartan guiones.

# %%
import re

# 1) Edad + sexo juntos: "60-year-old male", "57 year old woman", con opcional coma/descriptores
age_sex_re = re.compile(
    r"(?i)\b([0-9]{1,3})\s*[- ]?year[- ]?old\b[ ,;:\-]*\s*(male|female|man|woman)\b"
)

# 2) "male patient" / "female patient"
patient_sex_re = re.compile(r"(?i)\b(male|female)\s+patient\b")

# 3) Palabra suelta male/female/man/woman, evitando cosas como "female-to-male"
general_sex_re = re.compile(r"(?i)(?<!-)\b(male|female|man|woman)\b(?!-)")

def extract_sex_from_text_v2(note_text: str):
    """
    Devuelve:
      - sex_from_text_v2 (string original encontrado)
      - sex_from_text_norm_v2 ('male'|'female'), mapeando 'man'->'male' y 'woman'->'female'
    En orden de fiabilidad: age_sex_re > patient_sex_re > general_sex_re
    """
    t = (note_text or "").replace("\u2009"," ").replace("\u00a0"," ")
    for rx in (age_sex_re, patient_sex_re, general_sex_re):
        m = rx.search(t)
        if m:
            token = m.group(1).lower() if rx is not age_sex_re else m.group(2).lower()
            if token in {"man", "male"}:
                return token, "male"
            if token in {"woman", "female"}:
                return token, "female"
    return None, None

# Prueba rápida en 2 textos (opcional)
try:
    for txt in df["patient"].head(2):
        print(extract_sex_from_text_v2(txt))
except Exception as e:
    print("Prueba rápida: omito por error:", e)

# %% [markdown]
# CELDA 7A — DESPUÉS
# Ya tenemos un extractor mejorado para el SEXO.
# En la siguiente celda lo aplicamos a TODO el dataset y medimos `sex_match`.


('male', 'male')
('man', 'male')


Como puede observarse para los dos primeras notas se consigue extraer el gender

In [25]:
# %% [markdown]
# 🔗 CELDA 6-ter — ANTES
# Objetivo: Combinar la extracción de edad (la tuya) con el sexo v2 en una sola función.

# %%
def extract_age_sex_from_text_v2(note_text: str):
    t = (note_text or "").replace("\u2009"," ").replace("\u00a0"," ")
    out = {}

    # Edad (igual que tu celda)
    m = age_re_1.search(t) or age_re_2.search(t)
    out["age_from_text"] = int(m.group(1)) if m else None

    # Sexo (usa v2 + tu normalizador)
    tok, tok_norm = extract_sex_from_text_v2(t)
    out["sex_from_text"] = tok
    out["sex_from_text_norm"] = tok_norm

    return out

# Prueba rápida:
try:
    for txt in df["patient"].head(2):
        print(extract_age_sex_from_text_v2(txt))
except Exception as e:
    print("Prueba rápida: omito por error:", e)

# %% [markdown]
# CELDA 6-ter — DESPUÉS
# Ya tienes una única función con edad + sexo (v2).


{'age_from_text': 60, 'sex_from_text': 'male', 'sex_from_text_norm': 'male'}
{'age_from_text': 39, 'sex_from_text': 'man', 'sex_from_text_norm': 'male'}


Ahora extrae el año y el género  del texto de forma correcta para las dos primeras notas del dataset.

El siguiente paso será ver en qué porcentaje notas o filas conseguimos extraer el sexo y la edad desde el texto. De los casos en los que se extrae sexo y edad del texto, ¿Cuántos coinciden con las columnas sexo y género respectivamente?

In [27]:
# CELDA 7B — ANTES
# Objetivo: Evaluar **edad y sexo** extraídos desde el texto `patient` frente a las columnas normalizadas del CSV.
# Métricas que calculamos:
#   - Cobertura de extracción desde texto (qué % de filas lo traen).
#   - Exactitud (match) ignorando NaN: proporción de coincidencias cuando hay dato en ambos lados.
# Notas:
#   - Para SEXO: si existe extract_sex_from_text_v3() (con pronombres), lo usaremos; si no, usamos v2.
#   - Para EDAD: usamos los patrones age_re_1 / age_re_2 ya definidos arriba.

# %%
import pandas as pd
import numpy as np
import time
import re

# --- 1) Asegurar funciones/patrones para EDAD ---
# Si por lo que sea no existen en el entorno, los definimos aquí de forma segura.
if 'age_re_1' not in globals():
    age_re_1 = re.compile(r'(?i)\b([0-9]{1,3})-year-old\b')
if 'age_re_2' not in globals():
    age_re_2 = re.compile(r'(?i)\bage\s*[:=]?\s*([0-9]{1,3})\b')

def extract_age_from_text(note_text: str):
    """Devuelve edad (int) si encuentra 'X-year-old' o 'age: X'; en caso contrario None."""
    t = (note_text or "").replace("\u2009"," ").replace("\u00a0"," ")
    m = age_re_1.search(t) or age_re_2.search(t)
    if not m:
        return None
    try:
        return int(m.group(1))
    except:
        return None

# --- 2) Envoltorio para SEXO: usa v3 si existe; si no, v2 ---
def get_sex_from_text(note_text: str):
    """
    Devuelve tupla estandarizada:
      (sex_token, sex_norm, method)
      - sex_token: texto encontrado ('male','man', 'pronoun-male', etc.) o None
      - sex_norm : 'male'/'female' o None
      - method   : 'explicit' (v2) o 'pronoun_fallback'/'explicit' (v3)
    """
    t = (note_text or "").replace("\u2009"," ").replace("\u00a0"," ")
    # v3 disponible (pronombres)
    if 'extract_sex_from_text_v3' in globals():
        tok, tok_norm, method = extract_sex_from_text_v3(t)
        return tok, tok_norm, method
    # si no hay v3, usamos v2 (solo explícito)
    elif 'extract_sex_from_text_v2' in globals():
        tok, tok_norm = extract_sex_from_text_v2(t)
        method = 'explicit' if tok_norm else None
        return tok, tok_norm, method
    else:
        # seguridad por si faltara todo
        return None, None, None

# --- 3) Bucle por TODO el dataset ---
t0 = time.time()
rows = []

for r in df.itertuples(index=False):
    pid          = getattr(r, "patient_id", "")
    text         = getattr(r, "patient", "")
    age_col_norm = getattr(r, "age_col_norm", None)
    sex_col_norm = getattr(r, "sex_col_norm", None)

    # Extraer EDAD desde texto
    age_txt = extract_age_from_text(text)

    # Extraer SEXO desde texto (v3 si existe, si no v2)
    sex_tok, sex_norm, sex_method = get_sex_from_text(text)

    rows.append({
        "patient_id": pid,
        # Ground truth (CSV normalizado)
        "age_col_norm": age_col_norm,
        "sex_col_norm": sex_col_norm,
        # Texto
        "age_from_text": age_txt,
        "sex_from_text": sex_tok,
        "sex_from_text_norm": sex_norm,
        "sex_method": sex_method,
    })

cmp_df = pd.DataFrame(rows)

# --- 4) Tipos numéricos para EDAD (comparación limpia) ---
cmp_df["age_col_norm_int"]  = pd.to_numeric(cmp_df["age_col_norm"], errors="coerce").astype("Int64")
cmp_df["age_from_text_int"] = pd.to_numeric(cmp_df["age_from_text"], errors="coerce").astype("Int64")

# --- 5) Métricas de COBERTURA desde texto ---
age_cov_text = cmp_df["age_from_text_int"].notna().mean()                 # % de filas con edad extraída del texto
sex_cov_text = cmp_df["sex_from_text_norm"].notna().mean()                # % de filas con sexo extraído del texto

# --- 6) Métricas de EXACTITUD (match) ignorando NaN ---
cmp_df["age_match"] = np.where(
    cmp_df["age_from_text_int"].notna() & cmp_df["age_col_norm_int"].notna(),
    cmp_df["age_from_text_int"] == cmp_df["age_col_norm_int"],
    np.nan
)
cmp_df["sex_match"] = np.where(
    cmp_df["sex_from_text_norm"].notna() & cmp_df["sex_col_norm"].notna(),
    cmp_df["sex_from_text_norm"] == cmp_df["sex_col_norm"],
    np.nan
)

age_acc = cmp_df["age_match"].mean(skipna=True)  # exactitud condicional de edad
sex_acc = cmp_df["sex_match"].mean(skipna=True)  # exactitud condicional de sexo

t1 = time.time()

# --- 7) Reporte resumido ---
print(f"Filas procesadas: {len(cmp_df)}   Tiempo: {t1 - t0:.1f}s")
print(f"COBERTURA desde texto -> edad: {age_cov_text:.2%} | sexo: {sex_cov_text:.2%}")
print(f"EXACTITUD (match, ignora NaN) -> edad: {age_acc:.2%} | sexo: {sex_acc:.2%}")

# Muestra algunas filas para inspección
cols_show = [
    "patient_id",
    "age_col_norm_int","age_from_text_int","age_match",
    "sex_col_norm","sex_from_text","sex_from_text_norm","sex_method","sex_match"
]
try:
    display(cmp_df[cols_show].head(10))
except NameError:
    print(cmp_df[cols_show].head(10))


Filas procesadas: 167034   Tiempo: 15.8s
COBERTURA desde texto -> edad: 86.10% | sexo: 88.19%
EXACTITUD (match, ignora NaN) -> edad: 99.11% | sexo: 99.56%


patient_id  age_col_norm_int  age_from_text_int age_match sex_col_norm  \
0          0                60                 60      True         male   
1          1                39                 39      True         male   
2          2                57                 57      True         male   
3          3                69                 69      True         male   
4          4                57                 57      True         male   
5          5                52                 52      True         male   
6          6                59                 59      True       female   
7          7                33                 33      True       female   
8          8                66                 66      True         male   
9          9                66                 66      True         male   

  sex_from_text sex_from_text_norm sex_method  sex_match  
0          male               male   explicit        1.0  
1           man               male   explicit        1.0  
2          male               male   explicit        1.0  
3          male               male   explicit        1.0  
4          male               male   explicit        1.0  
5          male               male   explicit        1.0  
6        female             female   explicit        1.0  
7        female             female   explicit        1.0  
8          male               male   explicit        1.0  
9          male               male   explicit        1.0


En el siguiente paso vamos a convertir altura a metros y a normalizar los decimales (70,5 -> 70.5)


In [28]:
# CELDA 8A — ANTES
# Objetivo: utilidades de conversión de altura a metros y normalización de decimales.

# %%
import re

def _norm_decimals(s: str) -> str:
    """Convierte '70,5' -> '70.5' solo cuando la coma separa dígitos."""
    return re.sub(r'(?<=\d),(?=\d)', '.', s or "")

def height_to_m(value, unit):
    """Convierte altura (m/cm/mm/in/ft) a metros. Devuelve float o None."""
    try:
        v = float(value)
    except:
        return None
    u = (unit or "").lower().strip()
    if u in {"m","meter","meters"}:
        return v
    if u in {"cm","centimeter","centimeters"}:
        return v / 100.0
    if u in {"mm","millimeter","millimeters"}:
        return v / 1000.0
    if u in {"in","inch","inches"}:
        return v * 0.0254
    if u in {"ft","foot","feet"}:
        return v * 0.3048
    return None

def feet_inches_to_m(feet, inches=0):
    """Convierte pies+pulgadas a metros (e.g., 5'11\" -> m)."""
    try:
        f = float(feet); i = float(inches or 0)
    except:
        return None
    return f * 0.3048 + i * 0.0254

# CELDA 8A — DESPUÉS
# Listas las utilidades de altura. Seguimos con la extracción desde `patient`.


Extracción de Altura desde patient.

Regex para muchos formatos:

height/ht: 172 cm, ht 1.72 m

170 cm tall, 1.70 m tall

5'11" o 5 ft 11 in (o solo pies)

stature/body height ... 170 cm

Normalizaremos a metros y guardamos source para trazabilidad



In [29]:
# CELDA 8B — ANTES
# Objetivo: Extraer altura con múltiples patrones y normalizar a metros.

# %%
# Patrones "métricos" etiquetados
height_metric_re = re.compile(
    r'(?i)\b(?:height|ht)\b\s*[:=]?\s*([0-9]+(?:[.,][0-9]+)?)\s*(cm|centimeters?|m|meters?|mm|millimeters?)\b'
)

# "170 cm tall" / "1.70 m tall"
height_tall_re = re.compile(
    r'(?i)\b([0-9]+(?:[.,][0-9]+)?)\s*(cm|centimeters?|m|meters?)\s+tall\b'
)

# Contexto "stature" / "body height"
height_context_re = re.compile(
    r'(?i)\b(?:stature|body\s*height)\b[^0-9]{0,20}([0-9]+(?:[.,][0-9]+)?)\s*(cm|centimeters?|m|meters?|mm|millimeters?)\b'
)

# Pies/pulgadas: 5'11"  (permite 4–7 pies, 0–11 pulgadas)
height_ftin_quote_re = re.compile(
    r'(?i)\b([4-7])\'\s*([0-9]{1,2})(?:"|\'\')?'
)

# Pies/pulgadas con palabras: 5 ft 11 in
height_ftin_words_re = re.compile(
    r'(?i)\b([4-7])\s*(?:ft|feet|foot)\s*([0-9]{1,2})\s*(?:in|inch|inches)\b'
)

# Solo pies: 6 ft
height_ft_only_re = re.compile(
    r'(?i)\b([4-7])\s*(?:ft|feet|foot)\b'
)

def extract_height_from_text(note_text: str):
    """
    Devuelve dict con:
      - height_m (float)
      - height_raw (traza: value/unit)
      - height_source (patrón que lo detectó)
    """
    t = _norm_decimals((note_text or "")).replace("\u2009"," ").replace("\u00a0"," ")

    # Métrico etiquetado
    m = height_metric_re.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": float(val.replace(',', '.')), "unit": unit}, "height_source": "metric_tag"}

    # "cm/m tall"
    m = height_tall_re.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": float(val.replace(',', '.')), "unit": unit}, "height_source": "metric_tall"}

    # Contexto "stature/body height"
    m = height_context_re.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": float(val.replace(',', '.')), "unit": unit}, "height_source": "metric_context"}

    # Pies + pulgadas (con comillas)
    m = height_ftin_quote_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": f"{m.group(1)}'{m.group(2)}\"", "unit": "ft_in"}, "height_source": "ft_in_quote"}

    # Pies + pulgadas (con palabras)
    m = height_ftin_words_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": f"{m.group(1)} ft {m.group(2)} in", "unit": "ft_in"}, "height_source": "ft_in_words"}

    # Solo pies
    m = height_ft_only_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), 0)
        if hm is not None:
            return {"height_m": hm, "height_raw": {"value": f"{m.group(1)} ft", "unit": "ft"}, "height_source": "ft_only"}

    return {}

# Prueba rápida (3 textos)
try:
    for txt in df["patient"].head(3):
        print(extract_height_from_text(txt))
except Exception as e:
    print("Prueba rápida omitida:", e)

# CELDA 8B# El extractor devuelve metros + traza + source. Si no hay altura en el texto, devuelve {}.


{}
{}
{}


Vamos a ver cuantas notas mas o menos mencionan la altura ya que puede que el regex este mal o que simplemente en las tres primeras notas no se menciona nada de la altura:

In [30]:
# CELDA 8c# Objetivo: Estimar el "techo" de cobertura de ALTURA buscando pistas (cm/m/ft/in, height/ht/stature/length...).

# %%
import numpy as np
import pandas as pd
import re

pt = df["patient"].fillna("").str.lower()

cues = {
    "height/ht":  pt.str.contains(r'\b(height|ht)\b'),
    "stature":    pt.str.contains(r'\bstature\b'),
    "body height":pt.str.contains(r'\bbody\s*height\b'),
    "length":     pt.str.contains(r'\b(?:body\s*)?length\b'),  # neonatos a veces usan 'length'
    "cm":         pt.str.contains(r'\b[1-2]?\d{2}\s*cm\b'),    # 50–299 cm
    "meters m":   pt.str.contains(r'\b[1-2](?:[.,]\d{1,2})\s*m\b'),  # 1.4–1.99 m, etc.
    "ft/in words":pt.str.contains(r'\b[4-7]\s*(?:ft|feet|foot)\b'),
    "ft/in quotes":pt.str.contains(r'[4-7]\s*[\']\s*[0-9]{1,2}\s*(?:"|\'\')'),
    "inches":     pt.str.contains(r'\b[0-9]{1,2}\s*(?:in|inch|inches)\b'),
    "tall":       pt.str.contains(r'\btall\b'),
}

presence_any = np.zeros(len(df), dtype=bool)
for m in cues.values():
    presence_any |= m.fillna(False).to_numpy()

print("Pistas de ALTURA (% sobre total):")
for k, m in cues.items():
    print(f"- {k:<12}: {m.mean():.2%}")
print(f"\nAlguna pista de altura presente: {presence_any.mean():.2%}")

# Muestra 5 ejemplos con pistas (para inspección rápida)
idxs = np.flatnonzero(presence_any)[:5]
for i in idxs:
    print(f"\n--- patient_id={df.iloc[i]['patient_id']} ---")
    print(df.iloc[i]["patient"][:300].replace("\n"," "))


C:\Users\hered\AppData\Local\Temp\ipykernel_26116\3030981054.py:13: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  "height/ht":  pt.str.contains(r'\b(height|ht)\b'),


Pistas de ALTURA (% sobre total):
- height/ht   : 5.09%
- stature     : 0.62%
- body height : 0.08%
- length      : 5.45%
- cm          : 11.21%
- meters m    : 0.44%
- ft/in words : 0.09%
- ft/in quotes: 0.01%
- inches      : 5.04%
- tall        : 0.75%

Alguna pista de altura presente: 21.77%

--- patient_id=18 ---
A 67-year-old Caucasian female presented to our hospital with a chief complaint of persistent bright red blood per rectum. Her medical history was significant for hypertension, hyperlipidemia, diabetes mellitus type 2, coronary artery disease with three prior myocardial infarctions, recurrent cerebr

--- patient_id=22 ---
A 64-year-old Caucasian male smoker with a horseshoe kidney with a history of open pyelolithotomy 18 years ago, presented to King Abdulaziz Medical City in mid-2020 with a report from another hospital stating that he developed gross hematuria six months prior, which was treated as a urinary tract in

--- patient_id=23 ---
A 47-year-old Caucasian male with

Vamos a medir la cobertura real en Todo el dataset

In [32]:
# ALTURA — FIX RÁPIDO
# Objetivo: definir extract_height_from_text_v2 (y helpers si faltan) y calcular cobertura en TODO el dataset.

# %%
import re
import pandas as pd
import time

# --- Helpers (solo se definen si faltan en el entorno) ---
if '_norm_decimals' not in globals():
    def _norm_decimals(s: str) -> str:
        """Convierte '170,5' -> '170.5' solo si la coma separa dígitos."""
        return re.sub(r'(?<=\d),(?=\d)', '.', s or "")

if 'height_to_m' not in globals():
    def height_to_m(value, unit):
        """Convierte m/cm/mm/in/ft a metros."""
        try:
            v = float(value)
        except:
            return None
        u = (unit or "").lower().strip()
        if u in {"m","meter","meters"}:
            return v
        if u in {"cm","centimeter","centimeters","cms"}:
            return v/100.0
        if u in {"mm","millimeter","millimeters"}:
            return v/1000.0
        if u in {"in","inch","inches"}:
            return v*0.0254
        if u in {"ft","foot","feet"}:
            return v*0.3048
        return None

if 'feet_inches_to_m' not in globals():
    def feet_inches_to_m(feet, inches=0):
        """Convierte pies+pulgadas a metros."""
        try:
            f = float(feet); i = float(inches or 0)
        except:
            return None
        return f*0.3048 + i*0.0254

# --- Extractor ALTURA v2 ---
def extract_height_from_text_v2(note_text: str):
    """
    Devuelve:
      - height_m (float) en metros
      - height_raw: {'value': ..., 'unit': ...}
      - height_source: etiqueta del patrón que lo detectó
    """
    t = _norm_decimals((note_text or "")).replace("\u2009"," ").replace("\u00a0"," ")

    # 1) height/ht etiquetado con unidad
    height_label_v2 = re.compile(
        r'''(?ix)
        \b(?:height|ht|altura|estatura|talla)\b
        \s*(?:\([^)]+\))?\s*[:=]?\s*
        ([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
        '''
    )
    m = height_label_v2.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": float(val.replace(',', '.')), "unit": unit},
                "height_source": "label"
            }

    # 2) "height of/was/is/measured (at)/measuring 170 cm"
    height_of_was_measured = re.compile(
        r'''(?ix)
        \b(?:height|ht)\b
        [^0-9]{0,20}
        (?:of|was|is|measured\s*(?:at)?|measuring|at)\s*
        ([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
        '''
    )
    m = height_of_was_measured.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": float(val.replace(',', '.')), "unit": unit},
                "height_source": "of_was_measured"
            }

    # 3) Contextos "stature/body height/(body) length"
    height_context_v2 = re.compile(
        r'''(?ix)
        \b(?:stature|body\s*height|(?:body\s*)?length)\b
        [^0-9]{0,20}
        ([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
        '''
    )
    m = height_context_v2.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": float(val.replace(',', '.')), "unit": unit},
                "height_source": "context"
            }

    # 4) "170 cm tall" / "1.70 m tall"
    height_tall_v2 = re.compile(
        r'''(?ix)
        \b([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?)\s+tall\b
        '''
    )
    m = height_tall_v2.search(t)
    if m:
        val, unit = m.group(1), m.group(2)
        hm = height_to_m(val, unit)
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": float(val.replace(',', '.')), "unit": unit},
                "height_source": "tall"
            }

    # 5) Pies + pulgadas (comillas)
    height_ftin_quote_re = re.compile(
        r'''(?ix)\b([4-7])\s*(?:['\u2032])\s*([0-9]{1,2})\s*(?:"|\u2033|''|″)?'''
    )
    m = height_ftin_quote_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": f"{m.group(1)} ft {m.group(2)} in", "unit": "ft_in"},
                "height_source": "ft_in_quote"
            }

    # 6) Pies + pulgadas (palabras; admite coma)
    height_ftin_words_re = re.compile(
        r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\s*,?\s*([0-9]{1,2})\s*(?:in|inch|inches)\b'''
    )
    m = height_ftin_words_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": f"{m.group(1)} ft {m.group(2)} in", "unit": "ft_in"},
                "height_source": "ft_in_words"
            }

    # 7) Solo pies
    height_ft_only_re = re.compile(r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\b''')
    m = height_ft_only_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), 0)
        if hm is not None:
            return {
                "height_m": hm,
                "height_raw": {"value": f"{m.group(1)} ft", "unit": "ft"},
                "height_source": "ft_only"
            }

    return {}

# --- Aplicación a TODO el dataset (como querías) ---
t0 = time.time()
h_rows_v2 = []
for r in df.itertuples(index=False):
    pid  = getattr(r, "patient_id", "")
    text = getattr(r, "patient", "")
    ext  = extract_height_from_text_v2(text)  # <- ya definido
    h_rows_v2.append({
        "patient_id": pid,
        "height_m": ext.get("height_m"),
        "height_raw": ext.get("height_raw"),
        "height_source": ext.get("height_source"),
    })

height_df_v2 = pd.DataFrame(h_rows_v2)
cov_height_v2 = height_df_v2["height_m"].notna().mean()
t1 = time.time()

print(f"Filas procesadas: {len(height_df_v2)}   Tiempo: {t1 - t0:.1f}s")
print(f"Cobertura ALTURA (v2): {cov_height_v2:.2%}")

print("\nEjemplos CON altura (v2, hasta 10):")
try:
    display(height_df_v2[height_df_v2["height_m"].notna()].head(10))
except NameError:
    print(height_df_v2[height_df_v2["height_m"].notna()].head(10))

print("\nEjemplos SIN altura (v2, hasta 10):")
try:
    display(height_df_v2[height_df_v2["height_m"].isna()].head(10))
except NameError:
    print(height_df_v2[height_df_v2["height_m"].isna()].head(10))


Filas procesadas: 167034   Tiempo: 99.6s
Cobertura ALTURA (v2): 4.58%

Ejemplos CON altura (v2, hasta 10):


patient_id  height_m                      height_raw    height_source
22          22     1.660  {'value': 166.0, 'unit': 'cm'}  of_was_measured
38          38     1.080  {'value': 108.0, 'unit': 'cm'}            label
72          72     1.750    {'value': 1.75, 'unit': 'm'}             tall
75          75     0.530   {'value': 53.0, 'unit': 'cm'}          context
98          98     0.530   {'value': 53.0, 'unit': 'cm'}          context
144        144     0.530   {'value': 53.0, 'unit': 'cm'}          context
152        152     1.349  {'value': 134.9, 'unit': 'cm'}            label
182        182     0.630   {'value': 63.0, 'unit': 'cm'}  of_was_measured
210        210     0.029   {'value': 29.0, 'unit': 'mm'}          context
239        239     1.150  {'value': 115.0, 'unit': 'cm'}  of_was_measured


Ejemplos SIN altura (v2, hasta 10):


patient_id  height_m height_raw height_source
0          0       NaN       None          None
1          1       NaN       None          None
2          2       NaN       None          None
3          3       NaN       None          None
4          4       NaN       None          None
5          5       NaN       None          None
6          6       NaN       None          None
7          7       NaN       None          None
8          8       NaN       None          None
9          9       NaN       None          None

In [33]:
# CELDA 8B+++ — ANTES
# Objetivo: Mejorar cobertura de ALTURA con más variantes ("in height", "standing height", separadores, etc.)

# %%
import re

# Helpers por si no están ya en el entorno
if '_norm_decimals' not in globals():
    def _norm_decimals(s: str) -> str:
        return re.sub(r'(?<=\d),(?=\d)', '.', s or "")
if 'height_to_m' not in globals():
    def height_to_m(value, unit):
        try: v = float(value)
        except: return None
        u = (unit or "").lower().strip()
        if u in {"m","meter","meters"}: return v
        if u in {"cm","centimeter","centimeters","cms"}: return v/100.0
        if u in {"mm","millimeter","millimeters"}: return v/1000.0
        if u in {"in","inch","inches"}: return v*0.0254
        if u in {"ft","foot","feet"}: return v*0.3048
        return None
if 'feet_inches_to_m' not in globals():
    def feet_inches_to_m(feet, inches=0):
        try: f = float(feet); i = float(inches or 0)
        except: return None
        return f*0.3048 + i*0.0254

# 1) Etiquetas con separadores variados y paréntesis: Height / Ht / Altura / Estatura / Talla
height_label_v3p = re.compile(
    r'''(?ix)
    \b(?:height|ht\.?|altura|estatura|talla)\b
    (?:\s*\([^)]+\))?                             # opcional (cm), (standing), etc.
    (?:\s*[:=,\-]|\s+)?                           # separadores más tolerantes
    (?:was|is|of|at)?\s*                          # verbos/relaciones opcionales
    ([0-9]+(?:[.,][0-9]+)?)\s*                    # valor
    (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
    [)\],.;:]?                                    # puntuación final opcional
    '''
)

# 2) "height ... of/was/is/measured (at)/measuring ... 170 cm" (más flexible)
height_of_was_meas_v3p = re.compile(
    r'''(?ix)
    \b(?:height|ht\.?)\b
    [^0-9]{0,30}
    (?:of|was|is|measured\s*(?:at)?|measuring|at)?\s*
    ([0-9]+(?:[.,][0-9]+)?)\s*
    (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
    [)\],.;:]?
    '''
)

# 3) "<valor><unidad> in height|in stature" (ej: "170 cm in height")
value_in_height_v3p = re.compile(
    r'''(?ix)
    \b([0-9]+(?:[.,][0-9]+)?)\s*
    (cm|centimeters?|cms|m|meters?)\s+
    in\s+(?:height|stature)\b
    '''
)

# 4) "standing height ... 170 cm"
standing_height_v3p = re.compile(
    r'''(?ix)
    \bstanding\s+height\b
    [^0-9]{0,20}
    ([0-9]+(?:[.,][0-9]+)?)\s*
    (cm|centimeters?|cms|m|meters?)\b
    '''
)

# 5) "tall" con cm/m o con inches (extensión)
height_tall_v3p = re.compile(
    r'''(?ix)
    \b([0-9]+(?:[.,][0-9]+)?)\s*
    (cm|centimeters?|cms|m|meters?|in|inch|inches)\s+tall\b
    '''
)

# 6) Pies + pulgadas (comillas ASCII o tipográficas)
height_ftin_quotes_v3p = re.compile(
    r'''(?ix)\b([4-7])\s*(?:['\u2032])\s*([0-9]{1,2})\s*(?:"|\u2033|''|″)?'''
)
# 7) Pies + pulgadas con palabras (permite coma)
height_ftin_words_v3p = re.compile(
    r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\s*,?\s*([0-9]{1,2})\s*(?:in|inch|inches)\b'''
)
# 8) Solo pies (fallback)
height_ft_only_v3p = re.compile(r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\b''')

def extract_height_from_text_v3p(note_text: str):
    """
    Devuelve:
      - height_m (float)
      - height_raw: {'value': ..., 'unit': ...}
      - height_source: patrón que lo detectó
    """
    t = _norm_decimals((note_text or "")).replace("\u2009"," ").replace("\u00a0"," ")

    for rx, source in (
        (height_label_v3p, "label"),
        (height_of_was_meas_v3p, "of_was_measured"),
        (value_in_height_v3p, "value_in_height"),
        (standing_height_v3p, "standing_height"),
        (height_tall_v3p, "tall"),
    ):
        m = rx.search(t)
        if m:
            val, unit = m.group(1), m.group(2)
            hm = height_to_m(val, unit)
            if hm is not None:
                return {
                    "height_m": hm,
                    "height_raw": {"value": float(val.replace(',', '.')), "unit": unit},
                    "height_source": source
                }

    # Pies + pulgadas
    m = height_ftin_quotes_v3p.search(t) or height_ftin_words_v3p.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None:
            raw = f"{m.group(1)} ft {m.group(2)} in"
            return {"height_m": hm, "height_raw": {"value": raw, "unit": "ft_in"}, "height_source": "ft_in"}

    # Solo pies
    m = height_ft_only_v3p.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), 0)
        if hm is not None:
            raw = f"{m.group(1)} ft"
            return {"height_m": hm, "height_raw": {"value": raw, "unit": "ft"}, "height_source": "ft_only"}

    return {}
# CELDA 8B+++ — DESPUÉS
# Nuevo extractor v3+. A continuación lo aplicamos al dataset y comparamos cobertura con v2.


In [34]:
# CELDA 8C — ANTES
# Objetivo: Comparar coberturas de ALTURA entre v2 y v3+ y guardar la mejor como baseline.

# %%
import pandas as pd
import time

def _apply_height_extractor(extractor_fn):
    rows = []
    for r in df.itertuples(index=False):
        pid  = getattr(r, "patient_id", "")
        text = getattr(r, "patient", "")
        ext  = extractor_fn(text)
        rows.append({
            "patient_id": pid,
            "height_m": ext.get("height_m"),
            "height_raw": ext.get("height_raw"),
            "height_source": ext.get("height_source"),
        })
    out = pd.DataFrame(rows)
    return out, out["height_m"].notna().mean()

t0 = time.time()
height_df_v2b, cov_v2 = _apply_height_extractor(extract_height_from_text_v2)
height_df_v3p, cov_v3p = _apply_height_extractor(extract_height_from_text_v3p)
t1 = time.time()

print(f"Tiempo total: {t1 - t0:.1f}s")
print(f"Cobertura ALTURA v2  : {cov_v2:.2%}")
print(f"Cobertura ALTURA v3+: {cov_v3p:.2%}")

best = "v3+" if cov_v3p >= cov_v2 else "v2"
best_df = height_df_v3p if cov_v3p >= cov_v2 else height_df_v2b
best_df.to_csv("baseline_height_all.csv", index=False, encoding="utf-8")
print(f"Usaremos: {best}  | Guardado: baseline_height_all.csv")

print("\nEjemplos CON altura (mejor extractor, hasta 10):")
try:
    display(best_df[best_df["height_m"].notna()].head(10))
except NameError:
    print(best_df[best_df["height_m"].notna()].head(10))

# CELDA 8C — DESPUÉS
# - Si v3+ sube la cobertura notablemente, lo adoptamos.
# - Si aún queda muy por debajo del 21.77% (techo de "pistas"), puede que muchas 'pistas' de "cm/inches"
#   no sean altura humana (tumores, tamaños de lesiones, etc.). Si ves casos de altura que no captura,
#   pásame 2–3 textos y ajusto patrones específicos.


Tiempo total: 185.2s
Cobertura ALTURA v2  : 4.58%
Cobertura ALTURA v3+: 3.79%
Usaremos: v2  | Guardado: baseline_height_all.csv

Ejemplos CON altura (mejor extractor, hasta 10):


patient_id  height_m                      height_raw    height_source
22          22     1.660  {'value': 166.0, 'unit': 'cm'}  of_was_measured
38          38     1.080  {'value': 108.0, 'unit': 'cm'}            label
72          72     1.750    {'value': 1.75, 'unit': 'm'}             tall
75          75     0.530   {'value': 53.0, 'unit': 'cm'}          context
98          98     0.530   {'value': 53.0, 'unit': 'cm'}          context
144        144     0.530   {'value': 53.0, 'unit': 'cm'}          context
152        152     1.349  {'value': 134.9, 'unit': 'cm'}            label
182        182     0.630   {'value': 63.0, 'unit': 'cm'}  of_was_measured
210        210     0.029   {'value': 29.0, 'unit': 'mm'}          context
239        239     1.150  {'value': 115.0, 'unit': 'cm'}  of_was_measured

In [35]:
# CELDA 8B-tweak — ANTES
# Objetivo: endurecer el contexto y plausibilidad en la extracción de ALTURA v2.

# %%
import re

# Reutilizamos helpers existentes
def _plausible_height(value: float, unit: str) -> bool:
    u = (unit or "").lower().strip()
    try:
        v = float(value)
    except:
        return False
    if u in {"m","meter","meters"}:
        return 0.3 <= v <= 2.5
    if u in {"cm","centimeter","centimeters","cms"}:
        return 30 <= v <= 250
    if u in {"mm","millimeter","millimeters"}:
        return v >= 300   # >= 30 cm
    if u in {"in","inch","inches"}:
        # no se usa sola para altura normalmente, pero si aparece en "tall" lo convertimos
        return 12 <= v <= 100
    if u in {"ft","foot","feet"}:
        return 4 <= v <= 7
    return False

# Sustituimos SOLO el patrón de "contexto" y añadimos chequeo de plausibilidad en extract_height_from_text_v2
height_context_v2_strict = re.compile(
    r'''(?ix)
    \b(?:stature|body\s*height|body\s*length|crown(?:-\s*|\s*)(?:heel|rump)\s*length)\b
    [^0-9]{0,20}
    ([0-9]+(?:[.,][0-9]+)?)\s*
    (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b
    '''
)

def extract_height_from_text_v2(note_text: str):
    t = _norm_decimals((note_text or "")).replace("\u2009"," ").replace("\u00a0"," ")

    # 1) Etiqueta height/ht/... con unidad
    height_label_v2 = re.compile(
        r'''(?ix)\b(?:height|ht|altura|estatura|talla)\b
        \s*(?:\([^)]+\))?\s*[:=]?\s*
        ([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b'''
    )
    m = height_label_v2.search(t)
    if m and _plausible_height(m.group(1), m.group(2)):
        hm = height_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm,
                    "height_raw": {"value": float(m.group(1).replace(',', '.')), "unit": m.group(2)},
                    "height_source": "label"}

    # 2) "height ... of/was/is/measured (at)/measuring ... 170 cm"
    height_of_was_measured = re.compile(
        r'''(?ix)\b(?:height|ht)\b[^0-9]{0,20}
        (?:of|was|is|measured\s*(?:at)?|measuring|at)?\s*
        ([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|mm|millimeters?)\b'''
    )
    m = height_of_was_measured.search(t)
    if m and _plausible_height(m.group(1), m.group(2)):
        hm = height_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm,
                    "height_raw": {"value": float(m.group(1).replace(',', '.')), "unit": m.group(2)},
                    "height_source": "of_was_measured"}

    # 3) Contexto estricto (sin "length" genérico)
    m = height_context_v2_strict.search(t)
    if m and _plausible_height(m.group(1), m.group(2)):
        hm = height_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm,
                    "height_raw": {"value": float(m.group(1).replace(',', '.')), "unit": m.group(2)},
                    "height_source": "context"}

    # 4) "170 cm tall" / "1.70 m tall"
    height_tall_v2 = re.compile(
        r'''(?ix)\b([0-9]+(?:[.,][0-9]+)?)\s*
        (cm|centimeters?|cms|m|meters?|in|inch|inches)\s+tall\b'''
    )
    m = height_tall_v2.search(t)
    if m and _plausible_height(m.group(1), m.group(2)):
        hm = height_to_m(m.group(1), m.group(2))
        if hm is not None:
            return {"height_m": hm,
                    "height_raw": {"value": float(m.group(1).replace(',', '.')), "unit": m.group(2)},
                    "height_source": "tall"}

    # 5) Pies + pulgadas (cotas ya impuestas por el patrón)
    height_ftin_quote_re = re.compile(r'''(?ix)\b([4-7])\s*(?:['\u2032])\s*([0-9]{1,2})\s*(?:"|\u2033|''|″)?''')
    m = height_ftin_quote_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None and 0.3 <= hm <= 2.5:
            return {"height_m": hm,
                    "height_raw": {"value": f"{m.group(1)} ft {m.group(2)} in", "unit": "ft_in"},
                    "height_source": "ft_in_quote"}

    height_ftin_words_re = re.compile(r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\s*,?\s*([0-9]{1,2})\s*(?:in|inch|inches)\b''')
    m = height_ftin_words_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), m.group(2))
        if hm is not None and 0.3 <= hm <= 2.5:
            return {"height_m": hm,
                    "height_raw": {"value": f"{m.group(1)} ft {m.group(2)} in", "unit": "ft_in"},
                    "height_source": "ft_in_words"}

    # 6) Solo pies
    height_ft_only_re = re.compile(r'''(?ix)\b([4-7])\s*(?:ft|feet|foot)\b''')
    m = height_ft_only_re.search(t)
    if m:
        hm = feet_inches_to_m(m.group(1), 0)
        if hm is not None and 0.3 <= hm <= 2.5:
            return {"height_m": hm,
                    "height_raw": {"value": f"{m.group(1)} ft", "unit": "ft"},
                    "height_source": "ft_only"}

    return {}

# Recalcular cobertura tras el tweak (opcional)
h_rows_tweak = []
for r in df.itertuples(index=False):
    pid  = getattr(r, "patient_id", "")
    text = getattr(r, "patient", "")
    ext  = extract_height_from_text_v2(text)
    h_rows_tweak.append({"patient_id": pid, "height_m": ext.get("height_m"),
                         "height_raw": ext.get("height_raw"), "height_source": ext.get("height_source")})
_height_df_tweak = pd.DataFrame(h_rows_tweak)
print("Cobertura ALTURA (v2-tweak):", f"{_height_df_tweak['height_m'].notna().mean():.2%}")


Cobertura ALTURA (v2-tweak): 3.40%


La cobertura sigue siendo baja, vamos a recorrer todas las notas en busca de pistas de altura y lo vamos a comparar con el CSV que se generó

In [36]:
#ALTURA: pistas vs extracción
# =========================
import re, pandas as pd, numpy as np

# --- RUTAS (¡ajusta estas dos a tu entorno!)
DATA_CSV = r"C:/Users/hered/Desktop/TFM/TFM/data/PMC-Patients.csv"
HEIGHT_BASELINE_CSV = r"C:/Users/hered/Desktop/TFM/TFM/baseline_height_all.csv"   # el CSV que se generó con height_m, height_source, etc.

# --- Cargar dataset principal
df = pd.read_csv(DATA_CSV, dtype=str, encoding="utf-8", na_filter=False)
if "patient_id" not in df.columns:
    # Si no hay patient_id, crea uno incremental
    df["patient_id"] = np.arange(len(df)).astype(str)

# --- Cargar baseline de altura (el que generaste)
hb = pd.read_csv(HEIGHT_BASELINE_CSV, dtype=str, encoding="utf-8", na_filter=False)
# Normalizar tipos
hb["height_m"] = pd.to_numeric(hb.get("height_m", np.nan), errors="coerce")
hb["has_extracted_height"] = hb["height_m"].notna()

# --- Unir por patient_id
merged = df.merge(hb[["patient_id","height_m","height_raw","height_source","has_extracted_height"]],
                  on="patient_id", how="left")

# --- Normalizar texto
def norm_text(s: str) -> str:
    s = (s or "")
    s = s.replace("\u2009"," ").replace("\u00a0"," ")
    return s

merged["patient_norm"] = merged["patient"].apply(norm_text)

# Definir PISTAS de altura (cues)
# Nota: mantenemos expresiones “laxas” para estimar techo de cobertura
cues_patterns = {
    "label_height": re.compile(r"(?i)\b(height|ht\.?|altura|estatura|talla)\b"),
    "stature":      re.compile(r"(?i)\bstature\b"),
    "standing_height": re.compile(r"(?i)\bstanding\s+height\b"),
    "body_height":  re.compile(r"(?i)\bbody\s*height\b"),
    "body_length":  re.compile(r"(?i)\b(?:body\s*)?length\b"),
    # valores métricos plausibles cerca de cm/m
    "cm_value":     re.compile(r"(?i)\b(?:[3-9]\d|1\d{2}|2[0-4]\d|250)\s*cm\b"),  # 30-250 cm
    "m_value":      re.compile(r"(?i)\b(?:0\.\d{1,2}|1(?:\.\d{1,2})?|2(?:\.\d{1,2})?)\s*m\b"),  # 0.x-2.x m
    # pies/pulgadas
    "ft_words":     re.compile(r"(?i)\b[4-7]\s*(?:ft|feet|foot)\b"),
    "ft_quotes":    re.compile(r"(?i)\b[4-7]\s*['\u2032]"),
    "in_words":     re.compile(r"(?i)\b[0-9]{1,2}\s*(?:in|inch|inches)\b"),
    # expresiones “tall”
    "tall_metric":  re.compile(r"(?i)\b[0-9]+(?:[.,][0-9]+)?\s*(?:cm|centimeters?|m|meters?)\s+tall\b"),
    "tall_inches":  re.compile(r"(?i)\b[0-9]{1,2}\s*(?:in|inch|inches)\s+tall\b")
}

def find_cues(text: str):
    text = text or ""
    hits = []
    for name, rx in cues_patterns.items():
        if rx.search(text):
            hits.append(name)
    return hits

def first_snippets(text: str, max_hits=2, window=60):
    """Devuelve hasta 2 fragmentos alrededor de los primeros matches de cualquier cue."""
    text = text or ""
    spans = []
    for rx in cues_patterns.values():
        for m in rx.finditer(text):
            spans.append((m.start(), m.end()))
    spans.sort(key=lambda x: x[0])
    out = []
    for s,e in spans[:max_hits]:
        start = max(0, s - window)
        end   = min(len(text), e + window)
        out.append(text[start:end].replace("\n"," "))
    return out

merged["height_cues"] = merged["patient_norm"].apply(find_cues)
merged["has_height_cues"] = merged["height_cues"].apply(lambda L: len(L) > 0)
merged["height_cues_snippets"] = merged["patient_norm"].apply(first_snippets)

# --- Métricas globales
n = len(merged)
cue_rate = merged["has_height_cues"].mean()
extract_rate = merged["has_extracted_height"].fillna(False).mean()
both_rate = (merged["has_height_cues"] & merged["has_extracted_height"].fillna(False)).mean()

print(f"Filas: {n}")
print(f"- % con PISTAS de altura (cues): {cue_rate:.2%}")
print(f"- % con ALTURA extraída (baseline): {extract_rate:.2%}")
print(f"- % con ambas (pistas y extracción): {both_rate:.2%}")

# --- Tabla de frecuencias de cada cue
from collections import Counter
all_cues = Counter(c for L in merged["height_cues"] for c in L)
cue_freq = pd.DataFrame(sorted(all_cues.items(), key=lambda x: x[1], reverse=True),
                        columns=["cue","count"])
cue_freq["pct"] = (cue_freq["count"] / n * 100).round(2)
print("\nTop cues:")
print(cue_freq.head(20))

# --- Ejemplos a revisar: hay pistas pero NO se extrajo altura
to_review = merged[(merged["has_height_cues"]) & (~merged["has_extracted_height"].fillna(False))].copy()
cols_view = ["patient_id","height_cues","height_cues_snippets","patient_norm"]
print(f"\nEjemplos con PISTAS pero SIN extracción (muestra 10 de {len(to_review)}):")
display(to_review[cols_view].head(10))

# --- (Opcional) Ejemplos con extracción (para ver qué patrones sí capturamos)
ok_examples = merged[merged["has_extracted_height"].fillna(False)].copy()
ok_cols = ["patient_id","height_m","height_raw","height_source","height_cues","height_cues_snippets","patient_norm"]
print(f"\nEjemplos con extracción OK (muestra 10 de {len(ok_examples)}):")
display(ok_examples[ok_cols].head(10))

# --- Guardar informe completo para depurar offline
out_cols = ["patient_id","has_height_cues","height_cues","height_cues_snippets",
            "has_extracted_height","height_m","height_raw","height_source"]
report = merged[out_cols].copy()
report.to_csv("height_cues_vs_extraction.csv", index=False, encoding="utf-8")
print("\nGuardado informe: height_cues_vs_extraction.csv")

# --- Resumen rápido por tipos de cue (¿en cuáles fallamos más?)
fail = to_review.explode("height_cues")
fail_counts = (fail.groupby("height_cues")["patient_id"]
               .count()
               .sort_values(ascending=False)
               .rename("fails"))
summary = cue_freq.merge(fail_counts, left_on="cue", right_index=True, how="left").fillna(0)
summary["fail_rate_given_cue_%"] = (summary["fails"] / summary["count"] * 100).round(1)
print("\nCues con mayor tasa de fallo (pistas presentes pero sin extracción):")
display(summary.sort_values("fail_rate_given_cue_%", ascending=False).head(20))


Filas: 167034
- % con PISTAS de altura (cues): 17.66%
- % con ALTURA extraída (baseline): 4.58%
- % con ambas (pistas y extracción): 4.58%

Top cues:
                cue  count   pct
0          cm_value  10064  6.03
1       body_length   9109  5.45
2      label_height   8511  5.10
3          in_words   8424  5.04
4           m_value   1232  0.74
5           stature   1032  0.62
6       tall_metric    691  0.41
7         ft_quotes    609  0.36
8          ft_words    156  0.09
9       body_height    137  0.08
10  standing_height     72  0.04
11      tall_inches     29  0.02

Ejemplos con PISTAS pero SIN extracción (muestra 10 de 21844):


patient_id                         height_cues  \
18         18                       [body_length]   
23         23                          [in_words]   
27         27                          [in_words]   
41         41  [label_height, cm_value, in_words]   
42         42                          [in_words]   
43         43                          [cm_value]   
50         50                       [body_length]   
63         63                          [in_words]   
67         67                           [m_value]   
70         70                          [cm_value]   

                                 height_cues_snippets  \
18  [id colon. The mass was noncircumferential, me...   
23  [ll resolving. On presentation in 2016, visual...   
27  [ve bilaterally, and gait was ataxic. Strength...   
41  [d no fracture episodes, including fragility f...   
42  [ phenprocoumon was intensified to a new targe...   
43  [a mass on the right lobe of the cerebellum (2...   
50  [d soft tissue resulted in a large defect of a...   
63  [sive ophthalmology examination. He had a visu...   
67  [isease progression. Laboratory results showed...   
70  [5F) and a transseptal needle (standard Brocke...   

                                         patient_norm  
18  A 67-year-old Caucasian female presented to ou...  
23  A 47-year-old Caucasian male with a history of...  
27  A 59-year-old female, current smoker with 20 p...  
41  A 41-year-old Japanese woman presented to our ...  
42  A 14-year-old male adolescent presented to the...  
43  The patient is a 68-year-old retired male, bor...  
50  The patient was a 31-year-old male, a right-ha...  
63  A 5-year-old boy initially presented with head...  
67  A 61-year-old male was diagnosed with de novo ...  
70  The first patient is a 53-year-old male with a...


Ejemplos con extracción OK (muestra 10 de 7651):


patient_id  height_m                      height_raw    height_source  \
22          22     1.660  {'value': 166.0, 'unit': 'cm'}  of_was_measured   
38          38     1.080  {'value': 108.0, 'unit': 'cm'}            label   
72          72     1.750    {'value': 1.75, 'unit': 'm'}             tall   
75          75     0.530   {'value': 53.0, 'unit': 'cm'}          context   
98          98     0.530   {'value': 53.0, 'unit': 'cm'}          context   
144        144     0.530   {'value': 53.0, 'unit': 'cm'}          context   
152        152     1.349  {'value': 134.9, 'unit': 'cm'}            label   
182        182     0.630   {'value': 63.0, 'unit': 'cm'}  of_was_measured   
210        210     0.029   {'value': 29.0, 'unit': 'mm'}          context   
239        239     1.150  {'value': 115.0, 'unit': 'cm'}  of_was_measured   

                               height_cues  \
22                [label_height, cm_value]   
38                [label_height, cm_value]   
72                  [m_value, tall_metric]   
75                 [body_length, cm_value]   
98                 [body_length, cm_value]   
144                [body_length, cm_value]   
152            [label_height, body_length]   
182               [label_height, cm_value]   
210                [body_length, in_words]   
239  [label_height, body_length, cm_value]   

                                  height_cues_snippets  \
22   [perature, 37.1℃. The weight of the patient wa...   
38   [mal blood pressure. Her weight was 17 kg (50t...   
72   [A 20-year-old Caucasian male (1.75 m tall and...   
75   [child was registered as a male. Birth weight ...   
98   [At admission, her weight was 3100 g (25th–50t...   
144  [ 51 and 85. Her birth weight was 3550 g (75th...   
152  [ because of the patient's small anthropometri...   
182  [ic diarrhea, poor growth, and abnormal hair. ...   
210  [h high myopia presented with decreased visual...   
239  [t they were reassured. Compared to her peers,...   

                                          patient_norm  
22   A 64-year-old Caucasian male smoker with a hor...  
38   A 53-month-old Sudanese female presented with ...  
72   A 20-year-old Caucasian male (1.75 m tall and ...  
75   A patient of Ukrainian origin (UKR29) was born...  
98   A 20-day-old girl was admitted to the neonatal...  
144  This 7-year-old girl is the only child of non-...  
152  A 12-year-old girl developed abdominal pain an...  
182  This is a case of 41-month-old infant girl who...  
210  A 48-year-old Japanese woman with high myopia ...  
239  We present a 5-year-old female child who was s...


Guardado informe: height_cues_vs_extraction.csv

Cues con mayor tasa de fallo (pistas presentes pero sin extracción):


cue  count   pct   fails  fail_rate_given_cue_%
3          in_words   8424  5.04  7926.0                   94.1
10  standing_height     72  0.04    57.0                   79.2
7         ft_quotes    609  0.36   481.0                   79.0
4           m_value   1232  0.74   855.0                   69.4
1       body_length   9109  5.45  5895.0                   64.7
0          cm_value  10064  6.03  5451.0                   54.2
5           stature   1032  0.62   509.0                   49.3
2      label_height   8511  5.10  4055.0                   47.6
9       body_height    137  0.08    50.0                   36.5
11      tall_inches     29  0.02     9.0                   31.0
6       tall_metric    691  0.41     0.0                    0.0
8          ft_words    156  0.09     0.0                    0.0